index.merge.scheduler.max_thread_count is not validated correctly, can break indexing

**Elasticsearch version**: 2.3.4 / 2.4.0

**Plugins installed**: kopf

**JVM version**:

version: "1.8.0_102",
vm_name: "OpenJDK 64-Bit Server VM",
vm_version: "25.102-b14"",
vm_vendor: "Oracle Corporation",

**OS version**: 
Debian 8.5 
Linux 8e884c0e66a2 4.4.15-moby # 1 SMP Thu Jul 28 22:03:07 UTC 2016 x86_64 GNU/Linux

**Description of the problem including expected versus actual behavior**:

ES arbitrarily accepts any integer for `index.merge.scheduler.max_thread_count` even if it exceeds internal limits which then causes various scheduled tasks to throw exceptions and eventually break indexing.

**Steps to reproduce**:

We experienced this on our production cluster but I was able to reproduce it using a couple of docker containers.

  1. Create an index:
  ```
POST /test_index
{
  "number_of_shards": 4,
  "number_of_replicas": 1
}

{
  "acknowledged": true
}
  ```

  2. Update the max thread count:
  ```
PUT /test_index/_settings
{
  "index.merge.scheduler.max_thread_count": 12
}

{
  "acknowledged": true
}
  ```

  3. Alternate creating documents ...
  ```
POST /test_index/test_type
{
   "name": "bob"
}

{
  "_index": "test_index",
  "_type": "test_type",
  "_id": "AVcH80xPYp3inMh6kFUL",
  "_version": 1,
  "_shards": {
    "total": 2,
    "successful": 2,
    "failed": 0
  },
  "created": true
}
  ```
  And optimizing the index (to force a flush):
  ```
GET /test_index/_optimize
  ```

  4. This will shortly lead to translog failures:
  ```
POST /test_index/test_type
{
   "name": "bob"
}

{
  "error": {
    "root_cause": [
      {
        "type": "create_failed_engine_exception",
        "reason": "Create failed for [test_type#AVcH7buoYp3inMh6kFT4]",
        "shard": "3",
        "index": "test_index"
      }
    ],
    "type": "create_failed_engine_exception",
    "reason": "Create failed for [test_type#AVcH7buoYp3inMh6kFT4]",
    "shard": "3",
    "index": "test_index",
    "caused_by": {
      "type": "already_closed_exception",
      "reason": "translog is already closed"
    }
  },
  "status": 500
}
  ```

  ```
GET /test_index/_optimize
{
  "_shards": {
    "total": 8,
    "successful": 2,
    "failed": 6,
    "failures": [
      {
        "shard": 3,
        "index": "test_index",
        "status": "INTERNAL_SERVER_ERROR",
        "reason": {
          "type": "flush_failed_engine_exception",
          "reason": "Flush failed",
          "shard": "3",
          "index": "test_index",
          "caused_by": {
            "type": "already_closed_exception",
            "reason": "translog is already closed"
          }
        }
      }
    ]
  }
}
  ```

**Provide logs (if relevant)**:

The invalid `max_thread_count` causes all sorts of exceptions:

Updating the index settings:
```
[2016-09-08 03:50:48,676][INFO ][index.shard              ] [Abe Brown] [test_index][3] updating [index.merge.scheduler.max_thread_count] from [7] to [12]
[2016-09-08 03:50:48,677][WARN ][index.settings           ] [Abe Brown] [test_index] failed to refresh settings for [org.elasticsearch.index.shard.IndexShard$ApplyRefreshSettings@27098528]
java.lang.IllegalArgumentException: maxThreadCount should be <= maxMergeCount (= 7)
	at org.apache.lucene.index.ConcurrentMergeScheduler.setMaxMergesAndThreads(ConcurrentMergeScheduler.java:154)
	at org.elasticsearch.index.engine.ElasticsearchConcurrentMergeScheduler.refreshConfig(ElasticsearchConcurrentMergeScheduler.java:176)
	at org.elasticsearch.index.engine.InternalEngine.onSettingsChanged(InternalEngine.java:1268)
	at org.elasticsearch.index.shard.IndexShard$ApplyRefreshSettings.onRefreshSettings(IndexShard.java:1325)
	at org.elasticsearch.index.settings.IndexSettingsService.refreshSettings(IndexSettingsService.java:54)
	at org.elasticsearch.indices.cluster.IndicesClusterStateService.applySettings(IndicesClusterStateService.java:341)
	at org.elasticsearch.indices.cluster.IndicesClusterStateService.clusterChanged(IndicesClusterStateService.java:180)
	at org.elasticsearch.cluster.service.InternalClusterService.runTasksForExecutor(InternalClusterService.java:610)
	at org.elasticsearch.cluster.service.InternalClusterService$UpdateTask.run(InternalClusterService.java:772)
	at org.elasticsearch.common.util.concurrent.PrioritizedEsThreadPoolExecutor$TieBreakingPrioritizedRunnable.runAndClean(PrioritizedEsThreadPoolExecutor.java:231)
	at org.elasticsearch.common.util.concurrent.PrioritizedEsThreadPoolExecutor$TieBreakingPrioritizedRunnable.run(PrioritizedEsThreadPoolExecutor.java:194)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:745)
```

Various errors seen while indexing and optimizing:
```
[2016-09-08 03:58:16,128][WARN ][index.shard              ] [Abe Brown] [test_index][3] Failed to perform scheduled engine refresh
java.lang.IllegalArgumentException: maxThreadCount should be <= maxMergeCount (= 7)
	at org.apache.lucene.index.ConcurrentMergeScheduler.setMaxMergesAndThreads(ConcurrentMergeScheduler.java:154)
	at org.elasticsearch.index.engine.ElasticsearchConcurrentMergeScheduler.refreshConfig(ElasticsearchConcurrentMergeScheduler.java:176)
	at org.elasticsearch.index.engine.InternalEngine.refresh(InternalEngine.java:685)
	at org.elasticsearch.index.shard.IndexShard.refresh(IndexShard.java:661)
	at org.elasticsearch.index.shard.IndexShard$EngineRefresher$1.run(IndexShard.java:1343)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:745)
[2016-09-08 03:58:36,218][WARN ][indices.memory           ] [Abe Brown] failed to set shard [test_index][1] index buffer to [24.7mb]
org.apache.lucene.store.AlreadyClosedException: translog [2] is already closed
	at org.elasticsearch.index.translog.TranslogWriter.ensureOpen(TranslogWriter.java:329)
	at org.elasticsearch.index.translog.BufferingTranslogWriter.updateBufferSize(BufferingTranslogWriter.java:143)
	at org.elasticsearch.index.translog.Translog.updateBuffer(Translog.java:406)
	at org.elasticsearch.index.shard.IndexShard.updateBufferSize(IndexShard.java:1164)
	at org.elasticsearch.indices.memory.IndexingMemoryController.updateShardBuffers(IndexingMemoryController.java:232)
	at org.elasticsearch.indices.memory.IndexingMemoryController$ShardsIndicesStatusChecker.run(IndexingMemoryController.java:286)
	at org.elasticsearch.threadpool.ThreadPool$ReschedulingRunnable.doRun(ThreadPool.java:1057)
	at org.elasticsearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:37)
	at org.elasticsearch.threadpool.ThreadPool$LoggingRunnable.run(ThreadPool.java:668)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$201(ScheduledThreadPoolExecutor.java:180)
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:293)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:745)
```

```
[2016-09-08 03:53:05,512][DEBUG][action.admin.indices.forcemerge] [Abe Brown] [indices:admin/optimize] failed to execute operation for shard [[test_index][1], node[6jC3WEUrQWifFvTn-twoBw], [P], v[3], s[STARTED], a[id=Fa-2IEiESai0wk2csElq0Q]]
[test_index][[test_index][1]] FlushFailedEngineException[Flush failed]; nested: IllegalStateException[already committing a translog with generation: 1];
	at org.elasticsearch.index.engine.InternalEngine.flush(InternalEngine.java:762)
	at org.elasticsearch.index.engine.InternalEngine.forceMerge(InternalEngine.java:863)
	at org.elasticsearch.index.shard.IndexShard.forceMerge(IndexShard.java:793)
	at org.elasticsearch.action.admin.indices.forcemerge.TransportForceMergeAction.shardOperation(TransportForceMergeAction.java:79)
	at org.elasticsearch.action.admin.indices.forcemerge.TransportForceMergeAction.shardOperation(TransportForceMergeAction.java:46)
	at org.elasticsearch.action.support.broadcast.node.TransportBroadcastByNodeAction$BroadcastByNodeTransportRequestHandler.onShardOperation(TransportBroadcastByNodeAction.java:436)
	at org.elasticsearch.action.support.broadcast.node.TransportBroadcastByNodeAction$BroadcastByNodeTransportRequestHandler.messageReceived(TransportBroadcastByNodeAction.java:415)
	at org.elasticsearch.action.support.broadcast.node.TransportBroadcastByNodeAction$BroadcastByNodeTransportRequestHandler.messageReceived(TransportBroadcastByNodeAction.java:402)
	at org.elasticsearch.transport.TransportRequestHandler.messageReceived(TransportRequestHandler.java:33)
	at org.elasticsearch.transport.RequestHandlerRegistry.processMessageReceived(RequestHandlerRegistry.java:77)
	at org.elasticsearch.transport.TransportService$4.doRun(TransportService.java:376)
	at org.elasticsearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:37)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:745)
Caused by: java.lang.IllegalStateException: already committing a translog with generation: 1
	at org.elasticsearch.index.translog.Translog.prepareCommit(Translog.java:1785)
	at org.elasticsearch.index.engine.InternalEngine.flush(InternalEngine.java:753)
	... 14 more
```

```
[2016-09-08 03:59:27,335][WARN ][rest.suppressed          ] path: //test_index/test_type, params: {index=test_index, type=test_type}
RemoteTransportException[[Abe Brown][172.19.0.2:9300][indices:data/write/index[p]]]; nested: CreateFailedEngineException[Create failed for [test_type#AVcH80pjYp3inMh6kFUK]]; nested: AlreadyClosedException[translog is already closed];
Caused by: [test_index][[test_index][1]] CreateFailedEngineException[Create failed for [test_type#AVcH80pjYp3inMh6kFUK]]; nested: AlreadyClosedException[translog is already closed];
	at org.elasticsearch.index.engine.InternalEngine.create(InternalEngine.java:351)
	at org.elasticsearch.index.shard.IndexShard.create(IndexShard.java:545)
	at org.elasticsearch.index.engine.Engine$Create.execute(Engine.java:810)
	at org.elasticsearch.action.index.TransportIndexAction.executeIndexRequestOnPrimary(TransportIndexAction.java:236)
	at org.elasticsearch.action.index.TransportIndexAction.shardOperationOnPrimary(TransportIndexAction.java:157)
	at org.elasticsearch.action.index.TransportIndexAction.shardOperationOnPrimary(TransportIndexAction.java:66)
	at org.elasticsearch.action.support.replication.TransportReplicationAction$PrimaryPhase.doRun(TransportReplicationAction.java:648)
	at org.elasticsearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:37)
	at org.elasticsearch.action.support.replication.TransportReplicationAction$PrimaryOperationTransportHandler.messageReceived(TransportReplicationAction.java:279)
	at org.elasticsearch.action.support.replication.TransportReplicationAction$PrimaryOperationTransportHandler.messageReceived(TransportReplicationAction.java:271)
	at org.elasticsearch.transport.RequestHandlerRegistry.processMessageReceived(RequestHandlerRegistry.java:77)
	at org.elasticsearch.transport.TransportService$4.doRun(TransportService.java:376)
	at org.elasticsearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:37)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:745)
Caused by: org.apache.lucene.store.AlreadyClosedException: translog is already closed
	at org.elasticsearch.index.translog.Translog.ensureOpen(Translog.java:1888)
	at org.elasticsearch.index.translog.Translog.add(Translog.java:544)
	at org.elasticsearch.index.engine.InternalEngine.innerCreateNoLock(InternalEngine.java:437)
	at org.elasticsearch.index.engine.InternalEngine.innerCreate(InternalEngine.java:375)
	at org.elasticsearch.index.engine.InternalEngine.create(InternalEngine.java:346)
	... 15 more
```

Interestingly, closing and reopening the index will restore things to working order _with_ the configured thread count. 

Returning the `max_thread_count` to 1 shows that it has increased the thread count above what was previously set:

```
PUT /test_index/_settings
{
  "index.merge.scheduler.max_thread_count": 1
}
```

```
[2016-09-08 04:14:22,483][INFO ][index.shard              ] [Abe Brown] [test_index][1] updating [index.merge.scheduler.max_thread_count] from [17] to [1]
[2016-09-08 04:14:22,484][INFO ][index.shard              ] [Abe Brown] [test_index][2] updating [index.merge.scheduler.max_thread_count] from [17] to [1]
[2016-09-08 04:14:22,484][INFO ][index.shard              ] [Abe Brown] [test_index][3] updating [index.merge.scheduler.max_thread_count] from [17] to [1]
[2016-09-08 04:14:22,484][INFO ][index.shard              ] [Abe Brown] [test_index][0] updating [index.merge.scheduler.max_thread_count] from [17] to [1]
```

--- a/core/src/main/java/org/elasticsearch/index/IndexSettings.java
+++ b/core/src/main/java/org/elasticsearch/index/IndexSettings.java
@@ -263,8 +263,9 @@ public final class IndexSettings {
         scopedSettings.addSettingsUpdateConsumer(MergePolicyConfig.INDEX_MERGE_POLICY_MAX_MERGED_SEGMENT_SETTING, mergePolicyConfig::setMaxMergedSegment);
         scopedSettings.addSettingsUpdateConsumer(MergePolicyConfig.INDEX_MERGE_POLICY_SEGMENTS_PER_TIER_SETTING, mergePolicyConfig::setSegmentsPerTier);
         scopedSettings.addSettingsUpdateConsumer(MergePolicyConfig.INDEX_MERGE_POLICY_RECLAIM_DELETES_WEIGHT_SETTING, mergePolicyConfig::setReclaimDeletesWeight);
-        scopedSettings.addSettingsUpdateConsumer(MergeSchedulerConfig.MAX_THREAD_COUNT_SETTING, mergeSchedulerConfig::setMaxThreadCount);
-        scopedSettings.addSettingsUpdateConsumer(MergeSchedulerConfig.MAX_MERGE_COUNT_SETTING, mergeSchedulerConfig::setMaxMergeCount);
+
+        scopedSettings.addSettingsUpdateConsumer(MergeSchedulerConfig.MAX_THREAD_COUNT_SETTING, MergeSchedulerConfig.MAX_MERGE_COUNT_SETTING,
+            mergeSchedulerConfig::setMaxThreadAndMergeCount);
         scopedSettings.addSettingsUpdateConsumer(MergeSchedulerConfig.AUTO_THROTTLE_SETTING, mergeSchedulerConfig::setAutoThrottle);
         scopedSettings.addSettingsUpdateConsumer(INDEX_TRANSLOG_DURABILITY_SETTING, this::setTranslogDurability);
         scopedSettings.addSettingsUpdateConsumer(INDEX_TTL_DISABLE_PURGE_SETTING, this::setTTLPurgeDisabled);

--- a/core/src/main/java/org/elasticsearch/index/MergeSchedulerConfig.java
+++ b/core/src/main/java/org/elasticsearch/index/MergeSchedulerConfig.java
@@ -69,13 +69,14 @@ public final class MergeSchedulerConfig {
     private volatile int maxMergeCount;
 
     MergeSchedulerConfig(IndexSettings indexSettings) {
-        maxThreadCount = indexSettings.getValue(MAX_THREAD_COUNT_SETTING);
-        maxMergeCount = indexSettings.getValue(MAX_MERGE_COUNT_SETTING);
+        setMaxThreadAndMergeCount(indexSettings.getValue(MAX_THREAD_COUNT_SETTING),
+            indexSettings.getValue(MAX_MERGE_COUNT_SETTING));
         this.autoThrottle = indexSettings.getValue(AUTO_THROTTLE_SETTING);
     }
 
     /**
      * Returns <code>true</code> iff auto throttle is enabled.
+     *
      * @see ConcurrentMergeScheduler#enableAutoIOThrottle()
      */
     public boolean isAutoThrottle() {
@@ -100,8 +101,19 @@ public final class MergeSchedulerConfig {
      * Expert: directly set the maximum number of merge threads and
      * simultaneous merges allowed.
      */
-    void setMaxThreadCount(int maxThreadCount) {
+    void setMaxThreadAndMergeCount(int maxThreadCount, int maxMergeCount) {
+        if (maxThreadCount < 1) {
+            throw new IllegalArgumentException("maxThreadCount should be at least 1");
+        }
+        if (maxMergeCount < 1) {
+            throw new IllegalArgumentException("maxMergeCount should be at least 1");
+        }
+        if (maxThreadCount > maxMergeCount) {
+            throw new IllegalArgumentException("maxThreadCount (= " + maxThreadCount +
+                ") should be <= maxMergeCount (= " + maxMergeCount + ")");
+        }
         this.maxThreadCount = maxThreadCount;
+        this.maxMergeCount = maxMergeCount;
     }
 
     /**
@@ -109,13 +121,5 @@ public final class MergeSchedulerConfig {
      */
     public int getMaxMergeCount() {
         return maxMergeCount;
-    }
-
-    /**
-     *
-     * Expert: set the maximum number of simultaneous merges allowed.
-     */
-    void setMaxMergeCount(int maxMergeCount) {
-        this.maxMergeCount = maxMergeCount;
     }
 }

--- a/core/src/test/java/org/elasticsearch/indices/settings/UpdateSettingsIT.java
+++ b/core/src/test/java/org/elasticsearch/indices/settings/UpdateSettingsIT.java
@@ -28,7 +28,9 @@ import org.apache.logging.log4j.core.filter.RegexFilter;
 import org.elasticsearch.action.admin.cluster.health.ClusterHealthResponse;
 import org.elasticsearch.action.admin.cluster.node.stats.NodeStats;
 import org.elasticsearch.action.admin.cluster.node.stats.NodesStatsResponse;
+import org.elasticsearch.action.admin.indices.create.CreateIndexRequestBuilder;
 import org.elasticsearch.action.admin.indices.settings.get.GetSettingsResponse;
+import org.elasticsearch.action.admin.indices.settings.put.UpdateSettingsRequestBuilder;
 import org.elasticsearch.cluster.metadata.IndexMetaData;
 import org.elasticsearch.common.Priority;
 import org.elasticsearch.common.logging.Loggers;
@@ -414,6 +416,53 @@ public class UpdateSettingsIT extends ESIntegTestCase {
         } finally {
             TestLoggers.removeAppender(rootLogger, mockAppender);
             Loggers.setLevel(rootLogger, savedLevel);
+        }
+    }
+
+    public void testInvalidMergeMaxThreadCount() throws IllegalAccessException {
+        CreateIndexRequestBuilder createBuilder = prepareCreate("test")
+            .setSettings(Settings.builder()
+                .put(IndexMetaData.SETTING_NUMBER_OF_SHARDS, "1")
+                .put(IndexMetaData.SETTING_NUMBER_OF_REPLICAS, "0")
+                .put(MergePolicyConfig.INDEX_MERGE_POLICY_MAX_MERGE_AT_ONCE_SETTING.getKey(), "2")
+                .put(MergePolicyConfig.INDEX_MERGE_POLICY_SEGMENTS_PER_TIER_SETTING.getKey(), "2")
+                .put(MergeSchedulerConfig.MAX_THREAD_COUNT_SETTING.getKey(), "100")
+                .put(MergeSchedulerConfig.MAX_MERGE_COUNT_SETTING.getKey(), "10")
+            );
+        IllegalArgumentException exc = expectThrows(IllegalArgumentException.class,
+            () -> createBuilder.get());
+        assertThat(exc.getMessage(), equalTo("maxThreadCount (= 100) should be <= maxMergeCount (= 10)"));
+
+        assertAcked(prepareCreate("test")
+            .setSettings(Settings.builder()
+                .put(IndexMetaData.SETTING_NUMBER_OF_SHARDS, "1")
+                .put(IndexMetaData.SETTING_NUMBER_OF_REPLICAS, "0")
+                .put(MergePolicyConfig.INDEX_MERGE_POLICY_MAX_MERGE_AT_ONCE_SETTING.getKey(), "2")
+                .put(MergePolicyConfig.INDEX_MERGE_POLICY_SEGMENTS_PER_TIER_SETTING.getKey(), "2")
+                .put(MergeSchedulerConfig.MAX_THREAD_COUNT_SETTING.getKey(), "100")
+                .put(MergeSchedulerConfig.MAX_MERGE_COUNT_SETTING.getKey(), "100")
+            ));
+
+        {
+            UpdateSettingsRequestBuilder updateBuilder = client().admin().indices()
+                .prepareUpdateSettings("test")
+                .setSettings(Settings.builder()
+                    .put(MergeSchedulerConfig.MAX_THREAD_COUNT_SETTING.getKey(), "1000")
+                );
+            exc = expectThrows(IllegalArgumentException.class,
+                () -> updateBuilder.get());
+            assertThat(exc.getMessage(), equalTo("maxThreadCount (= 1000) should be <= maxMergeCount (= 100)"));
+        }
+
+        {
+            UpdateSettingsRequestBuilder updateBuilder = client().admin().indices()
+                .prepareUpdateSettings("test")
+                .setSettings(Settings.builder()
+                    .put(MergeSchedulerConfig.MAX_MERGE_COUNT_SETTING.getKey(), "10")
+                );
+            exc = expectThrows(IllegalArgumentException.class,
+                () -> updateBuilder.get());
+            assertThat(exc.getMessage(), equalTo("maxThreadCount (= 100) should be <= maxMergeCount (= 10)"));
         }
     }
 

